### PR TITLE
Add tests for PDT in JIT

### PR DESCRIPTION
**Pull request by nikithamalgifb**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#56029 Add tests for PDT in JIT**
* #55376 Update script API to take example inputs

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -13440,6 +13440,16 @@ def call(self):
         self.assertTrue(isinstance(f, torch.jit.ScriptModule))
         self.assertTrue(isinstance(f.call(), property))
 
+    def test_monkeytype(self):
```
Comment (gmagogsfm) [inline test/test_jit.py:13443]:
```suggestion
    def test_pdt(self):
```

File: test/jit/test_monkeytype.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import torch
+
+def test_sum_monkey(a, b):
```
Comment (gmagogsfm) [inline test/jit/test_monkeytype.py:3]: Really excited to see these working.

File: test/test_jit.py [hunk 3]
```diff
@@ -13440,6 +13440,16 @@ def call(self):
...
+        scripted_fn_add = torch.jit._script_pdt(test_sum_monkey, example_inputs=[(3, 4)])
+        scripted_fn_sub = torch.jit._script_pdt(test_sub_monkey, example_inputs=[(3.9, 4.10)])
+        scripted_fn_mul = torch.jit._script_pdt(test_mul_monkey, example_inputs=[(-10, 9)])
+        scripted_fn_complex = torch.jit._script_pdt(test_args_complex, example_inputs=[(torch.rand(3, 4), torch.rand(3, 4))])
+        self.assertEqual(scripted_fn_add(10, 2), test_sum_monkey(10, 2))
+        self.assertEqual(scripted_fn_sub(6.5, 2.9), test_sub_monkey(6.5, 2.9))
+        self.assertEqual(scripted_fn_mul(-1, 3), test_mul_monkey(-1, 3))
+        arg1, arg2 = torch.rand(3, 4), torch.rand(3, 4)
+        self.assertEqual(scripted_fn_complex(arg1, arg2), test_args_complex(arg1, arg2))
```
Comment (gmagogsfm) [inline test/test_jit.py:13452]: This complex function doesn't seem like a good test case because `test_args_complex` takes two tensors, which would work without PDT.

File: test/jit/test_monkeytype.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import torch
```
Comment (gmagogsfm) [inline test/jit/test_monkeytype.py:1]: `monkeytype` is another library, the feature's name is PDT, please rename file.

File: test/test_jit.py (hunk 1 above)
Comment (gmagogsfm) [inline test/test_jit.py:13443]: Why isn't this test case in same file as those functions being tested?

File: test/jit/test_monkeytype.py (hunk 2 above)
Comment (gmagogsfm) [inline test/jit/test_monkeytype.py:3]: These functions can be nested functions inside test case, then use `make_global` to expose them to global name space. That would provide better readability and isolation.

Comment (gmagogsfm) [inline test/jit/test_monkeytype.py:3]: Consider adding more tests covering non-numeric types, like bool, str, tuple, list

File: test/jit/test_monkeytype.py (hunk 4 above)
Comment (nikithamalgifb) [reply test/jit/test_monkeytype.py]: Agreed and fixed

File: test/jit/test_monkeytype.py (hunk 2 above)
Comment (nikithamalgifb) [reply test/jit/test_monkeytype.py]: Agreed. Will add

File: test/test_jit.py (hunk 1 above)
Comment (nikithamalgifb) [reply test/test_jit.py]: Now moved.

Comment (nikithamalgifb) [reply test/test_jit.py]: Fixed now

File: test/jit/test_pdt.py [hunk 5]
```diff
@@ -42,6 +42,12 @@ def test_str(a):
 def test_list_and_tuple(a):
     return sum(a)
 
+def test_dict(a):
```
Comment (gmagogsfm) [inline test/jit/test_pdt.py:45]: Does current implementation work for `nn.Module` methods and normal class methods? If so, could we add some test cases?

File: test/jit/test_pdt.py [hunk 6]
```diff
@@ -2,7 +2,7 @@
 import sys
 import torch
 from torch.testing._internal.jit_utils import JitTestCase
-from typing import List
+from typing import List  # noqa F401
```
[the remark this answers is not in the document]
Comment (nikithamalgifb) [reply test/jit/test_pdt.py]: During Runtime, TorchScript fails to identify `List` and I run into `RuntimeError` . Hence, I need to add this.

File: test/jit/test_monkeytype.py (hunk 2 above)
Comment (nikithamalgifb) [reply test/jit/test_monkeytype.py]: I have moved everything to `test_pdt.py`. So I guess this is not required

File: test/test_jit.py (hunk 3 above)
Comment (nikithamalgifb) [reply test/test_jit.py]: I agree. But I ended finding a bug in PDT while annotating types. Hence, I thought it can stay :)

File: test/jit/test_pdt.py (hunk 5 above)
Comment (nikithamalgifb) [reply test/jit/test_pdt.py:50]: Tests for `nn.Module` methods are still pending. I'm working on it

File: test/jit/test_monkeytype.py (hunk 2 above)
Comment (nikithamalgifb) [reply test/jit/test_monkeytype.py]: Fixed now.

File: torch/jit/_script.py [hunk 7]
```diff
@@ -44,8 +44,8 @@
 
 type_trace_db = JitTypeTraceStore()  # DB to hold all call traces from MonkeyType
 
-torch._C.ScriptMethod.graph_for = _graph_for  # type: ignore
-torch._C.ScriptFunction.graph_for = _graph_for  # type: ignore
+torch._C.ScriptMethod.graph_for = _graph_for  # type: ignore[attr-defined]
+torch._C.ScriptFunction.graph_for = _graph_for  # type: ignore[attr-defined]
```
Comment (gmagogsfm) [inline torch/jit/_script.py:48]: Can we make another PR for this only? This change is unrelated to the feature being added.

File: test/jit/test_pdt.py [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+        scripted_fn_int = torch.jit._script_pdt(test_list_and_tuple, example_inputs=[((3, 4, 5), )])
+        self.assertEqual(scripted_fn_int((1, 2, 3)), test_list_and_tuple((1, 2, 3)))
+
+    def test_pdt_dict(self):
+        def test_dict(a):
+            return a['foo']
+
+        make_global(test_dict)
+
+        _input = {'foo' : True, 'bar': False}
```
Comment (gmagogsfm) [inline test/jit/test_pdt.py:102]:
```suggestion
        inp = {'foo' : True, 'bar': False}
```

Comment (gmagogsfm) [reply test/jit/test_pdt.py:102]: This is same issue as we talked about in last PR, please avoid using underscore incorrectly.

File: test/jit/test_pdt.py [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+from torch.testing._internal.jit_utils import JitTestCase, make_global
+from torch.jit._monkeytype_config import _IS_MONKEYTYPE_INSTALLED
+from typing import List, Dict, Tuple  # noqa: F401
+
+# Make the helper files in test/ importable
+pytorch_test_dir = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
+sys.path.append(pytorch_test_dir)
+
+if not _IS_MONKEYTYPE_INSTALLED:
+    print("monkeytype is not installed. Skipping tests for Profile-Directed Typing ", file=sys.stderr)
```
Comment (gmagogsfm) [inline test/jit/test_pdt.py:13]:
```suggestion
    print("monkeytype is not installed. Skipping tests for Profile-Directed Typing", file=sys.stderr)
```

File: test/jit/test_pdt.py [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+
+        def test_str(a):
+            if a == "":
+                return False
+            else:
+                return True
+
+        make_global(test_sum, test_sub, test_mul, test_args_complex, test_bool, test_str)
+
+        scripted_fn_add = torch.jit._script_pdt(test_sum, example_inputs=[(3, 4)])
```
Comment (gmagogsfm) [inline test/jit/test_pdt.py:54]: Please move definition of function to use of function.

```
def test_sum():
    ......

scripted_fn_sum = torch.jit._script_pdt(test_sum)
self.assertEqual(scripted_fn_sum(), fn_sum())
```

File: test/jit/test_pdt.py [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+    JitTestCase = object  # type: ignore[misc, assignment] # noqa: F811
+
+if __name__ == "__main__":
+    raise RuntimeError(
+        "This test file is not meant to be run directly, use:\n\n"
+        "\tpython test/test_jit.py TESTNAME\n\n"
+        "instead."
+    )
+
+class TestPDT(JitTestCase):
```
Comment (gmagogsfm) [inline test/jit/test_pdt.py:23]: Have you tested whether using `torch.jit._script_pdt` as decorator on functions work?

File: test/jit/test_pdt.py [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+
+    def test_class_methods(self):
+        class M1:
+            def fn(a):  # noqa: B902
+                return a
+
+        make_global(M1)
+
+        scripted_fn = torch.jit._script_pdt(M1.fn, example_inputs=[(10, )])
+        self.assertEqual(scripted_fn(M1.fn(2)), M1.fn(2))  # type: ignore[arg-type]
```
Comment (gmagogsfm) [inline test/jit/test_pdt.py:151]: There seems to be quite some lint surppression, what are they about? Is there a good reason to have these lint offenses?

**Comment by gmagogsfm** [comment]
In the future, please bundle tests together with implementation. Otherwise, it is creating a small window (between two PRs land) where your feature isn't covered by tests.

File: test/jit/test_pdt.py (hunk 12 above)
Comment (nikithamalgifb) [reply test/jit/test_pdt.py:158]: Yes. So these throw the below errors:
```
error: Argument 1 to "fn" has incompatible type "int"; expected "M1"  [arg-type]
```
So it's expecting the first argument to be of type class `M1` for `fn`

File: test/jit/test_pdt.py (hunk 11 above)
Comment (nikithamalgifb) [reply test/jit/test_pdt.py:23]: No I haven't. Let me give it a try.

File: test/jit/test_pdt.py (hunk 12 above)
Comment (gmagogsfm) [reply test/jit/test_pdt.py:158]: I think this is because first argument of a class method should always be `self`, which must be typed as the class. So it is a real issue.

Lint warnings are generally useful at identifying underlying issues, suppressing them is sort of like removing all smoke detectors in a house : )

File: test/jit/test_pdt.py (hunk 11 above)
Comment (nikithamalgifb) [reply test/jit/test_pdt.py:23]: Created an issue for it.

File: test/jit/test_pdt.py (hunk 12 above)
Comment (nikithamalgifb) [reply test/jit/test_pdt.py:158]: I agree, but adding the self, I run into runtime error:
I don't think even in torch.jit.script this is not supported. 

Here is an example:

```
class example:
   def fn(self, b: int):
      return b

s = torch.jit.script(example.fn)
print(s(20))
```
Traceback (most recent call last):
  File "example.py", line 15, in <module>
    print(s(20))
RuntimeError: fn() Expected a value of type 'Tensor (inferred)' for argument 'self' but instead found type 'int'.
Inferred 'self' to be of type 'Tensor' because it was not annotated with an explicit type.
Position: 0
Value: 20
Declaration: fn(Tensor self, int b) -> (int)
Cast error details: Unable to cast Python instance of type <class 'int'> to C++ type 'at::Tensor'
```
Even this:
```
s = torch.jit.script(example)
print(s.fn(20))

Traceback (most recent call last):
  File "example.py", line 12, in <module>
    print(s.fn(20))
TypeError: fn() missing 1 required positional argument: 'b'

```

File: test/jit/test_pdt.py [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+
+class TestPDT(JitTestCase):
+    """
+    A suite of tests for profile directed typing in TorchScript.
+    """
+    def test_pdt(self):
+        def test_sum(a, b):
+            return a + b
+
+        make_global(test_sum)
```
Comment (gmagogsfm) [inline test/jit/test_pdt.py:31]: I think we don't need `make_global` for functions, just classes.

File: test/jit/test_pdt.py [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+            if a == "":
+                return False
+            else:
+                return True
+
+        make_global(test_str)
+        scripted_fn_str = torch.jit._script_pdt(test_str, example_inputs=[("",)])
+        self.assertEqual(scripted_fn_str("abc"), test_str("abc"))
+
+    def test_pdt_list_and_tuple(self):
```
Comment (gmagogsfm) [inline test/jit/test_pdt.py:77]: Could you try some nested data structure as well? like List[Tuple] or List[List]

File: test/jit/test_pdt.py [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+                                            ([3, 4, 5],), (True, ), ({"a": True}, ), ])
+        self.assertEqual(scripted_fn(10), test_multiple_type_refinement(10))
+        self.assertEqual(scripted_fn("def"), test_multiple_type_refinement("def"))
+        self.assertEqual(scripted_fn(7.89999), test_multiple_type_refinement(7.89999))
+        self.assertEqual(scripted_fn([10, 11, 14]), test_multiple_type_refinement([10, 11, 14]))
+        self.assertEqual(scripted_fn(False), test_multiple_type_refinement(False))
+        self.assertEqual(scripted_fn({"abc" : True, "def": False}), test_multiple_type_refinement({"abc" : True, "def": False}))
+
+    def test_class_as_profiled_types(self):
+        class UserDefinedClass:
```
Comment (gmagogsfm) [inline test/jit/test_pdt.py:154]: Nice!

File: test/jit/test_pdt.py [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+            def fn(self, b):
+                assert b is not None
```
Comment (gmagogsfm) [inline test/jit/test_pdt.py:156]: This statement doesn't really prove that `b`'s type is correctly inferred since all types can be compared against `None`.

**Review by gmagogsfm — CHANGES_REQUESTED** [review]

File: test/jit/test_pdt.py (hunk 13 above)
Comment (nikithamalgifb) [reply test/jit/test_pdt.py:31]: I tried removing global for tests, it did not work.

File: test/jit/test_pdt.py (hunk 16 above)
Comment (nikithamalgifb) [reply test/jit/test_pdt.py:156]: Fixed now.

File: test/jit/test_pdt.py [hunk 17]
```diff
@@ -90,3 +90,204 @@ def fun(self, x):
...
+        self.assertEqual(scripted_fn_float([11.9, 7.6]), test_list_and_tuple([11.9, 7.6]))
+
+        scripted_fn_bool = torch.jit._script_pdt(test_list_and_tuple, example_inputs=[([True, False, True],)])
+        self.assertEqual(scripted_fn_bool([True, True, True]), test_list_and_tuple([True, True, True]))
+
+        scripted_fn_int = torch.jit._script_pdt(test_list_and_tuple, example_inputs=[([3, 4, 5], )])
+        self.assertEqual(scripted_fn_int([1, 2, 3]), test_list_and_tuple([1, 2, 3]))
+
+        scripted_fn_float = torch.jit._script_pdt(test_list_and_tuple, example_inputs=[((4.9, 8.9),)])
+        self.assertEqual(scripted_fn_float((11.9, 7.6)), test_list_and_tuple((11.9, 7.6)))
```
Comment (gmagogsfm) [inline test/jit/test_pdt.py:159]: FYI @ansley 

These tests actually reminds me of a next step feature we need:

When combining types like `List[int]` and `List[str]`, maybe we can generate `List[Any]` instead of `Any`, so we retain a little more information about the type.

Comment (gmagogsfm) [reply test/jit/test_pdt.py:159]: Another thing I realized is that:
When we have generic function/method overloading support (@tugsbayasgalan is working on part of it now), maybe we don't need to combine the types any more in some cases, instead we just generate a few different overloads of method, this will be lower runtime overhead and can capture type relationships like `type(a) === type(b)` 

for example

```
def fn(a, b):
    return a + b
```

The intention of function may be that type of `a` and `b` are always the same. This is lost if we analyze types of both to be `Any`. However, this relationship will be kept if we generate two different overloads of `fn`:
```
def fn(a: int, b: int):
   ...

def fn(a: float, b: float):
  ...
```

File: test/jit/test_pdt.py [hunk 18]
```diff
@@ -90,3 +90,204 @@ def fun(self, x):
...
+        scripted_fn_str = torch.jit._script_pdt(test_str, example_inputs=[("",)])
+        self.assertEqual(scripted_fn_str("abc"), test_str("abc"))
+
+    def test_pdt_list_and_tuple(self):
+        def test_list_and_tuple(a):
+            return sum(a)
+
+        make_global(test_list_and_tuple)
+
+        scripted_fn_float = torch.jit._script_pdt(test_list_and_tuple, example_inputs=[([4.9, 8.9],)])
```
[the remark this answers is not in the document]
Comment (nikithamalgifb) [reply test/jit/test_pdt.py]: Those aren't copy paste errors, the first three are with `list` inputs and the rest are with `tuple` inputs. I have renamed them now.

File: test/jit/test_pdt.py (hunk 13 above)
Comment (nikithamalgifb) [reply test/jit/test_pdt.py:98]: I remember now why `make_global` is required. `monkeytype` actually fails to recognize the `test_sum` inside `test_pdt`. It only recognizes when it is made global. 

Here is the error:
```
Traceback (most recent call last):
  File "/data/users/nikithamalgi/pytorch/test/jit/test_pdt.py", line 100, in test_pdt
    self.assertEqual(scripted_fn_add(10, 2), test_sum(10, 2))
  File "/data/users/nikithamalgi/pytorch/torch/testing/_internal/common_utils.py", line 133, in prof_func_call
    return prof_callable(func_call, *args, **kwargs)
  File "/data/users/nikithamalgi/pytorch/torch/testing/_internal/common_utils.py", line 130, in prof_callable
    return callable(*args, **kwargs)
RuntimeError: test_sum() Expected a value of type 'Tensor (inferred)' for argument 'a' but instead found type 'int'.
Inferred 'a' to be of type 'Tensor' because it was not annotated with an explicit type.
Position: 0
Value: 10
Declaration: test_sum(Tensor a, Tensor b) -> (Tensor)
Cast error details: Unable to cast Python instance to C++ type (compile in debug mode for details)
```

Comment (gmagogsfm) [reply test/jit/test_pdt.py:98]: I see. Is this a fundamental limitation of MonkeyType? ie, MonkeyType can not profile nested functions?

We don't need to do anything about it in this PR, but if this is a limitation of MonkeyType, we would need keep it in mind.

**Review by gmagogsfm — APPROVED** [review]

**Comment by nikithamalgifb** [comment]
Opened a new PR #58211, since i am having issues importing this PR
